### DISCUSSION TITLE
Compaq Portable

### DISCUSSION THREAD

**Discussion by AtomicLunchbox** [640-KB/GLaBIOS#45] [Suggestions]
So I couldn't resist and I had to try GLaBIOS out on my Compaq Portable, probably better to discuss here than on a commit. Pain to get at the ROM sockets, but I'd say it's a success and pretty close to working.

![PXL_20240227_043951506](https://github.com/640-KB/GLaBIOS/assets/130497829/e880bfcd-c2c9-405e-8263-a505cac627f9)

My setup is the v1 board, with the original Compaq VDU, Compaq Floppy controller and an XTIDE. The two option roms are two copies of GLaTICK, but I don't actually have the RTC board in at the moment. It's also running a V20 with a PCSprint so this was booting at 7.16MHz.

The floppy error may be cosmetic, the drive seeks, is readable and bootable via the XTIDE menu though I didn't try to write anything.

It boots in the higher res 80 column mode and will switch via the mode command to regular 40 column CGA and back. I don't have anything except the mode command from Compaq DOS 3.31 to switch between the high and low frequency 80 column modes and that always stays in the MDA link mode no matter what commands I send. My guess is that an external CGA monitor won't work correctly, but I can try one out at another time.

Played a few games like Digger and Planet X3, ran some tests in CheckIt and didn't see anything out of the ordinary.

I didn't make any changes to commit 24dd103 beyond adding a extra beep when I was playing around in an emulate. The BIOS works better on real hardware than it does on 86box where the floppy doesn't work at all and, for some reason, digger remastered hangs?

Amazing work! I can't believe this is so close!

**Comment by 640-KB** [comment]
> So I couldn't resist and I had to try GLaBIOS out on my Compaq Portable, probably better to discuss here than on a commit. Pain to get at the ROM sockets, but I'd say it's a success and pretty close to working.

Thanks for doing it!

> My setup is the v1 board, with the original Compaq VDU, Compaq Floppy controller and an XTIDE. The two option roms are two copies of GLaTICK, but I don't actually have the RTC board in at the moment. It's also running a V20 with a PCSprint so this was booting at 7.16MHz.
> The floppy error may be cosmetic, the drive seeks, is readable and bootable via the XTIDE menu though I didn't try to write anything.

That's interesting about the floppy seek. The Compaq BIOS doesn't do any kind of seek test on POST right?   I'll have to do some research and see if there's anything known about the Compaq floppy controller/drives that's different than the PC.  In the meantime you could set `POST_TEST_FD = 0` which should just skip the seek test and not show an error.

In theory it's possible too that the PCSprint board running everything faster might push some of the floppy timing off a little bit.  If it seeks and reads though, then it's all working so that is interesting.  Could try commenting out the `TURBO_TYPE = TURBO_NONE` in the Compaq config section and see if that makes a difference.

Yeah, noticed GLaTICK appears to be loading twice. Just inadvertently put it in the image twice or are you using a 4K ROM that's wrapping addresses? Curious, which RTC board do you have?

> 
> It boots in the higher res 80 column mode and will switch via the mode command to regular 40 column CGA and back. I don't have anything except the mode command from Compaq DOS 3.31 to switch between the high and low frequency 80 column modes and that always stays in the MDA link mode no matter what commands I send. My guess is that an external CGA monitor won't work correctly, but I can try one out at another time.

Yeah any BIOS logic that handles a mode switch just doesn't exist yet. Right now, all it's doing is making up for the differences in system config DIP switches and the Compaq video CRTC register values. So all that's on the TODO list!

> I didn't make any changes to commit [24dd103](https://github.com/640-KB/GLaBIOS/commit/24dd103ce354ddd2e6f1a913d420f72853366fdf) beyond adding an extra beep when I was playing around in an emulate. 

Oh, because the Compaq beeps twice at boot and you just want it to sound familiar? That's a neat idea - we should add that!

> The BIOS works better on real hardware than it does on 86box where the floppy doesn't work at all and, for some reason, digger remastered hangs?
> 
> Amazing work! I can't believe this is so close!

Excellent!  This is all great information to have and really appreciate all of the testing and help.  If you run into anything else or even come across any specs or technical info about how it could work, please pass it along.

640KB

**Reply by AtomicLunchbox** [reply]
> That's interesting about the floppy seek. The Compaq BIOS doesn't do any kind of seek test on POST right? I'll have to do some research and see if there's anything known about the Compaq floppy controller/drives that's different than the PC. In the meantime you could set POST_TEST_FD = 0 which should just skip the seek test and not show an error.

I'll have to check the behavior, but it's such a pain to swap the BIOS ROM. I'll set POST_TEST_FD to 0 next time I build too. 

> In theory it's possible too that the PCSprint board running everything faster might push some of the floppy timing off a little bit. If it seeks and reads though, then it's all working so that is interesting. Could try commenting out the TURBO_TYPE = TURBO_NONE in the Compaq config section and see if that makes a difference.

I dialed it down to 4.77, because you are correct that the Compaq floppy controller is a bit flaky at the high speed, and had the same behavior. It does it on warm and cold boots too. I have another floppy controller on hand that I tried out, this one here: https://stason.org/TULARC/pc/io-cards/C-D/DTK-COMPUTER-INC-Multi-I-O-card-PII-147-HEXA-I-O-P.html and saw the same behavior. Clock on that works with GLaTICK too (and you're right, I have the image doubled up on a 4K ROM and have been too lazy to rewrite it).

>> It boots in the higher res 80 column mode and will switch via the mode command to regular 40 column CGA and back. I don't have anything except the mode command from Compaq DOS 3.31 to switch between the high and low frequency 80 column modes and that always stays in the MDA link mode no matter what commands I send. My guess is that an external CGA monitor won't work correctly, but I can try one out at another time.

> Yeah any BIOS logic that handles a mode switch just doesn't exist yet. Right now, all it's doing is making up for the differences in system config DIP switches and the Compaq video CRTC register values. So all that's on the TODO list!

I poked around in the code a bit more after I had posted and I think I understand it a bit better now. Should be easy enough to set the CRTC registers based on key strokes as a rough approximation of how the Compaq BIOS works, but I assume it must store this value somewhere too since it seems remember which flavor it was in when returning from CGA graphics mode. Maybe we can find a resource somewhere that documents this.

I'll monkey around with it and let you know if I get anywhere. It boots so fast though ... even where it's at today it'll be hard to go back the Compaq BIOS!

**Comment by 640-KB** [comment]
Another question - do you get the FDC error if you do a warm boot (ctrl-alt-del) or only cold?

**Comment by 640-KB** [comment]
> > That's interesting about the floppy seek. The Compaq BIOS doesn't do any kind of seek test on POST right? I'll have to do some research and see if there's anything known about the Compaq floppy controller/drives that's different than the PC. In the meantime you could set POST_TEST_FD = 0 which should just skip the seek test and not show an error.
> 
> I'll have to check the behavior, but it's such a pain to swap the BIOS ROM. I'll set POST_TEST_FD to 0 next time I build too.
 
Totally understand!  Don't need to keep changing it around... am just thinking out loud about things that might give a clue before having to tear into code and all.

> > In theory it's possible too that the PCSprint board running everything faster might push some of the floppy timing off a little bit. If it seeks and reads though, then it's all working so that is interesting. Could try commenting out the TURBO_TYPE = TURBO_NONE in the Compaq config section and see if that makes a difference.
> 
> I dialed it down to 4.77, because you are correct that the Compaq floppy controller is a bit flaky at the high speed, and had the same behavior. It does it on warm and cold boots too. I have another floppy controller on hand that I tried out, this one here: https://stason.org/TULARC/pc/io-cards/C-D/DTK-COMPUTER-INC-Multi-I-O-card-PII-147-HEXA-I-O-P.html and saw the same behavior. 

Huh... that is perplexing, especially if you are using another card and if the drive otherwise works properly in DOS.  Does it boot normally from a floppy with either controller?  The thing that causes that test to fail is if the FDC chip returns any non-success on a RECAL or SEEK operation (which are the effectively the same thing).  I could try making a debug build that outputs what the FDC is returning and might give a clue. Just strange because any read operation will of course do the same exact seek operation and would return a failure before even trying to read, so something is somehow different between POST and DOS... hmm.

The `TURBO_TYPE = TURBO_NONE` will actually make the build behave more like a Turbo PC and there is a part of the floppy code where it adds a larger I/O settling time for reading out the result from the FDC register stack.  For a future build you could try leaving the `TURBO_TYPE` the way it is and put in a higher value here, perhaps start with changing the 10 to 50 and see if that makes a difference:
https://github.com/640-KB/GLaBIOS/blob/e7f3697b4705da664b1e0d894980472cc7ab99ef/src/GLABIOS.ASM#L7785

> > > It boots in the higher res 80 column mode and will switch via the mode command to regular 40 column CGA and back. I don't have anything except the mode command from Compaq DOS 3.31 to switch between the high and low frequency 80 column modes and that always stays in the MDA link mode no matter what commands I send. My guess is that an external CGA monitor won't work correctly, but I can try one out at another time.
> 
> I poked around in the code a bit more after I had posted and I think I understand it a bit better now. Should be easy enough to set the CRTC registers based on key strokes as a rough approximation of how the Compaq BIOS works, but I assume it must store this value somewhere too since it seems remember which flavor it was in when returning from CGA graphics mode. Maybe we can find a resource somewhere that documents this.

That all makes sense - if I recall correctly the Compaq's unique video mode switching is all controller by special code the BIOS to support it.  I feel like I've read something somewhere that explained it in more detail - perhaps even enough to work from - but where that was escapes me now.  If you turn anything up first, send it along!

**Comment by AtomicLunchbox** [comment]
I tore it down and put the original BIOS (version C) back in. The floppy definitely does seek on boot with the Compaq BIOS. I'm going try to build an ISA ROM card that I can set to address F0000 so I can (hopefully) pull the BIOS ROMs off the board and have something easier to swap in and out. They really did put those ROMs in the worst possible place!

I found this reference that discusses the VDU: https://www.seasip.info/VintagePC/compaq.html and this one that documents some INT 10H extensions that look similar to what the Compaq DOS "mode" command can do: https://pdos.csail.mit.edu/6.828/2008/readings/hardware/vgadoc/COMPAQ.TXT

I'm well outside of the limits of knowledge here so not sure if this will help at all.

**Comment by AtomicLunchbox** [comment]
I have a pull request up to add the mode switching keyboard commands like the stock BIOS provides, but in the meantime I think I've figured out the floppy problems above.  I didn't read the code close enough and missed how some of the values are set - the Compaq Portable basically uses the same switch settings as the 5160 so doesn't need any handling. Gotta test on real hardware before I submit it.

**Reply by AtomicLunchbox** [reply]
This is wrong, it sets up the switches like a 5150 :)
